File: xlserver/xlSheet.py
import xlwings as xw
import xlGlobal as g

async def add(p):
    try:
        wb = g.xlApp.books(p["f"])
        n = p["n"] if "n" in p else None
        a = wb.sheets(p["a"]) if "a" in p else None
        b = wb.sheets(p["b"]) if "b" in p else None
        sh = wb.sheets.add(name = n, after = a, before = b)
        bi = g.getBookInfo(wb)
        bi["ns"] = { "n": sh.name, "i": sh.index }
        return bi
    except Exception as e:
        return g.errReturn()

async def setName(p):
  try:
      wb = g.xlApp.books(p["f"])
      sh = wb.sheets(p["i"])
      sh.name = p["n"]
  except Exception as e:
      return g.errReturn()

async def setTabColor(p):
  try:
      wb = g.xlApp.books(p["f"])
      sh = wb.sheets(p["i"])
      macro = g.xlApp.macro("tmkSetTabColor")
      macro(sh, *p["c"])
  except Exception as e:
      return g.errReturn()

async def resetTabColor(p):
  try:
      wb = g.xlApp.books(p["f"])
      sh = wb.sheets(p["i"])
      macro = g.xlApp.macro("tmkResetTabColor")
      macro(sh)
  except Exception as e:
      return g.errReturn()

async def getTabColor(p):
  try:
      wb = g.xlApp.books(p["f"])
      sh = wb.sheets(p["i"])
      macro = g.xlApp.macro("tmkGetTabColor")
      return macro(sh)
  except Exception as e:
      return g.errReturn()

async def activate(p):
  try:
      wb = g.xlApp.books(p["f"])
      sh = wb.sheets(p["i"])
      sh.activate()
  except Exception as e:
      return g.errReturn()

def addCallbacks():
    g.addCallback("xl", "sh", "as", add)
    g.addCallback("xl", "sh", "sn", setName)
    g.addCallback("xl", "sh", "stc", setTabColor)
    g.addCallback("xl", "sh", "rtc", resetTabColor)
    g.addCallback("xl", "sh", "gtc", getTabColor)
    g.addCallback("xl", "sh", "ac", activate)

File: runtime/excel/lib/xlSheet.js
import { ws, __ws } from './xlWebSocket.js';
import { ActiveWorkbook } from './xlWorkbook.js';
import { _Range } from './xlRange.js';

export class _Sheets {
    constructor(workbook) {
        this.workbook = workbook;
    }

    get Count() {
        return this.workbook.sheetsByIndex.length;
    }

    async add(p) {
        Blockly.checkStop();
        let param = { 'f': this.workbook.name };
        if (typeof p == 'string') {
            param.n = p;
        }
        else if (typeof p == 'object') {
            if (p.hasOwnProperty('name')) {
                param.n = p.name;
            }
            if (p.hasOwnProperty('before')) {
                param.b = p.before;
            }
            if (p.hasOwnProperty('after')) {
                param.a = p.after;
            }
        }
        let wb_data = await ws.send(__ws, "xl", "sh", "as", param);
        let sh_data = wb_data.ns;
        let sheet = new _Sheet(this.workbook, sh_data.i, sh_data.n);
        this.workbook.sheetsByName[sh_data.n] = sheet;
        this.workbook.sheetsByIndex.splice(sheet.index - 1, 0, sheet);
        for (let i = sheet.index; i < this.workbook.sheetsByIndex.length; i++) {
            this.workbook.sheetsByIndex[i].index++;
        }
        return sheet;
    }
}

export class _Sheet {
    constructor(workbook, index, name) {
        this.workbook = workbook;
        this.index = index;
        this.name = name;
    }

    get Parent() {
        return this.workbook;
    }

    get Name() {
        return this.name;
    }

    get Index() {
        return this.index;
    }

    Range(adrs) {
        let rng = new _Range(this.workbook, this, adrs, null, null, null, null);
        return rng;
    }

    Cells(r, c) {
        let rng = new _Range(this.workbook, this, null, r, c, null, null);
        return rng;
    }

    CellsRange(r1, c1, r2, c2) {
        let rng = new _Range(this.workbook, this, null, r1, c1, r2, c2);
        return rng;
    }

    async activate() {
        Blockly.checkStop();
        let param = { 'f': this.workbook.name, 'i': this.index };
        await ws.send(__ws, "xl", "sh", "ac", param);
        this.workbook.activeSheet = this;
    }

    async setName(name) {
        Blockly.checkStop();
        let oldName = this.name;
        let param = { 'f': this.workbook.name, 'i': this.index, 'n': name };
        await ws.send(__ws, "xl", "sh", "sn", param);
        this.name = name;
        delete this.workbook.sheetsByName[oldName];
        this.workbook.sheetsByName[name] = this;
    }

    async setTabColor(color) {
        Blockly.checkStop();
        let param = { 'f': this.workbook.name, 'i': this.index, 'c': color };
        await ws.send(__ws, "xl", "sh", "stc", param);
    }

    async resetTabColor() {
        Blockly.checkStop();
        let param = { 'f': this.workbook.name, 'i': this.index };
        await ws.send(__ws, "xl", "sh", "rtc", param);
    }

    async getTabColor() {
        Blockly.checkStop();
        let param = { 'f': this.workbook.name, 'i': this.index };
        return await ws.send(__ws, "xl", "sh", "gtc", param);
    }
}

export function Sheets(p) {
    return ActiveWorkbook().Sheets(p)
}

export function ActiveSheet() {
    return ActiveWorkbook().ActiveSheet();
}
